File: codeupipe/core/state.py
"""
State: Pipeline Execution Metadata

State tracks what happened during pipeline execution — which filters ran,
which were skipped, timing data, and errors encountered.
Access it after pipeline.run() via pipeline.state.
"""

from typing import Any, Dict, List, Optional, Tuple, Set

__all__ = ["State"]


class State:
    """
    Pipeline execution state — tracks filter execution, timing, and errors.

    Provides visibility into what happened during a pipeline run:
    - Which filters executed and in what order
    - Which filters were skipped (by valves)
    - Errors encountered during execution
    - Arbitrary metadata for custom tracking
    """

    def __init__(self):
        self.executed: List[str] = []
        self.skipped: List[str] = []
        self.errors: List[Tuple[str, Exception]] = []
        self.metadata: Dict[str, Any] = {}
        self.chunks_processed: Dict[str, int] = {}
        self.timings: Dict[str, float] = {}

    def mark_executed(self, name: str) -> None:
        """Record that a filter executed."""
        self.executed.append(name)

    def mark_skipped(self, name: str) -> None:
        """Record that a filter was skipped."""
        self.skipped.append(name)

    def increment_chunks(self, name: str, count: int = 1) -> None:
        """Increment the chunk counter for a streaming step."""
        self.chunks_processed[name] = self.chunks_processed.get(name, 0) + count

    def record_timing(self, name: str, duration: float) -> None:
        """Record step execution duration in seconds."""
        self.timings[name] = duration

    def record_error(self, name: str, error: Exception) -> None:
        """Record an error from a filter."""
        self.errors.append((name, error))

    def set(self, key: str, value: Any) -> None:
        """Store arbitrary metadata."""
        self.metadata[key] = value

    def get(self, key: str, default: Any = None) -> Any:
        """Retrieve metadata."""
        return self.metadata.get(key, default)

    @property
    def has_errors(self) -> bool:
        """Whether any errors were recorded."""
        return len(self.errors) > 0

    @property
    def last_error(self) -> Optional[Exception]:
        """The most recent error, or None."""
        return self.errors[-1][1] if self.errors else None

    def reset(self) -> None:
        """Reset state for a fresh run."""
        self.executed.clear()
        self.skipped.clear()
        self.errors.clear()
        self.metadata.clear()
        self.chunks_processed.clear()
        self.timings.clear()

    def diff(self, other: 'State') -> Dict[str, Any]:
        """Compare this state with another — what changed between runs."""
        result: Dict[str, Any] = {}

        added = [s for s in other.executed if s not in self.executed]
        removed = [s for s in self.executed if s not in other.executed]
        if added:
            result["added_steps"] = added
        if removed:
            result["removed_steps"] = removed

        timing_changes: Dict[str, Dict[str, Optional[float]]] = {}
        all_steps: Set[str] = set(self.timings) | set(other.timings)
        for step in sorted(all_steps):
            old_t = self.timings.get(step)
            new_t = other.timings.get(step)
            if old_t != new_t:
                timing_changes[step] = {"old": old_t, "new": new_t}
        if timing_changes:
            result["timing_changes"] = timing_changes

        old_errors: Set[str] = {name for name, _ in self.errors}
        new_errors: Set[str] = {name for name, _ in other.errors}
        if old_errors != new_errors:
            result["error_changes"] = {
                "added": sorted(new_errors - old_errors),
                "removed": sorted(old_errors - new_errors),
            }

        return result

    def __repr__(self) -> str:
        return (
            f"State(executed={self.executed}, skipped={self.skipped}, "
            f"errors={len(self.errors)}, timings={len(self.timings)}, "
            f"chunks={self.chunks_processed})"
        )

File: ports/go/codeupipe/state.go
package codeupipe

import (
	"fmt"
	"sort"
	"strings"
)

// State tracks what happened during pipeline execution — which filters ran,
// which were skipped, timing data, and errors encountered.
//
// Port of codeupipe/core/state.py
type State struct {
	Executed        []string
	Skipped         []string
	Errors          []StateError
	Metadata        map[string]any
	ChunksProcessed map[string]int
	Timings         map[string]float64
}

// StateError records a filter name and the error it produced.
type StateError struct {
	Name string
	Err  error
}

// NewState creates a fresh State.
func NewState() *State {
	return &State{
		Metadata:        make(map[string]any),
		ChunksProcessed: make(map[string]int),
		Timings:         make(map[string]float64),
	}
}

// MarkExecuted records that a filter executed.
func (s *State) MarkExecuted(name string) {
	s.Executed = append(s.Executed, name)
}

// MarkSkipped records that a filter was skipped.
func (s *State) MarkSkipped(name string) {
	s.Skipped = append(s.Skipped, name)
}

// IncrementChunks increments the chunk counter for a streaming step.
func (s *State) IncrementChunks(name string, count int) {
	s.ChunksProcessed[name] += count
}

// RecordTiming records step execution duration in seconds.
func (s *State) RecordTiming(name string, duration float64) {
	s.Timings[name] = duration
}

// RecordError records an error from a filter.
func (s *State) RecordError(name string, err error) {
	s.Errors = append(s.Errors, StateError{Name: name, Err: err})
}

// Set stores arbitrary metadata.
func (s *State) Set(key string, value any) {
	s.Metadata[key] = value
}

// Get retrieves metadata, with an optional default.
func (s *State) Get(key string, defaultValue ...any) any {
	if v, ok := s.Metadata[key]; ok {
		return v
	}
	if len(defaultValue) > 0 {
		return defaultValue[0]
	}
	return nil
}

// HasErrors returns whether any errors were recorded.
func (s *State) HasErrors() bool {
	return len(s.Errors) > 0
}

// LastError returns the most recent error, or nil.
func (s *State) LastError() error {
	if len(s.Errors) == 0 {
		return nil
	}
	return s.Errors[len(s.Errors)-1].Err
}

// Reset clears state for a fresh run.
func (s *State) Reset() {
	s.Executed = nil
	s.Skipped = nil
	s.Errors = nil
	s.Metadata = make(map[string]any)
	s.ChunksProcessed = make(map[string]int)
	s.Timings = make(map[string]float64)
}

// Diff compares this state with another — what changed between runs.
func (s *State) Diff(other *State) map[string]any {
	result := make(map[string]any)

	// Added / removed steps
	oldSet := toSet(s.Executed)
	newSet := toSet(other.Executed)

	var added, removed []string
	for _, name := range other.Executed {
		if !oldSet[name] {
			added = append(added, name)
		}
	}
	for _, name := range s.Executed {
		if !newSet[name] {
			removed = append(removed, name)
		}
	}
	if len(added) > 0 {
		result["added_steps"] = added
	}
	if len(removed) > 0 {
		result["removed_steps"] = removed
	}

	// Timing changes
	allSteps := make(map[string]bool)
	for k := range s.Timings {
		allSteps[k] = true
	}
	for k := range other.Timings {
		allSteps[k] = true
	}
	timingChanges := make(map[string]any)
	sorted := sortedKeys(allSteps)
	for _, step := range sorted {
		oldT, oldOk := s.Timings[step]
		newT, newOk := other.Timings[step]
		if oldOk != newOk || oldT != newT {
			tc := map[string]any{}
			if oldOk {
				tc["old"] = oldT
			}
			if newOk {
				tc["new"] = newT
			}
			timingChanges[step] = tc
		}
	}
	if len(timingChanges) > 0 {
		result["timing_changes"] = timingChanges
	}

	// Error changes
	oldErrors := make(map[string]bool)
	newErrors := make(map[string]bool)
	for _, e := range s.Errors {
		oldErrors[e.Name] = true
	}
	for _, e := range other.Errors {
		newErrors[e.Name] = true
	}
	var errAdded, errRemoved []string
	for name := range newErrors {
		if !oldErrors[name] {
			errAdded = append(errAdded, name)
		}
	}
	for name := range oldErrors {
		if !newErrors[name] {
			errRemoved = append(errRemoved, name)
		}
	}
	sort.Strings(errAdded)
	sort.Strings(errRemoved)
	if len(errAdded) > 0 || len(errRemoved) > 0 {
		result["error_changes"] = map[string]any{
			"added":   errAdded,
			"removed": errRemoved,
		}
	}

	return result
}

// String returns a human-readable representation.
func (s *State) String() string {
	return fmt.Sprintf(
		"State(executed=[%s], skipped=[%s], errors=%d, timings=%d)",
		strings.Join(s.Executed, ", "),
		strings.Join(s.Skipped, ", "),
		len(s.Errors),
		len(s.Timings),
	)
}

func toSet(items []string) map[string]bool {
	m := make(map[string]bool, len(items))
	for _, s := range items {
		m[s] = true
	}
	return m
}

func sortedKeys(m map[string]bool) []string {
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return keys
}
